app/metal-controller-manager/cmd/agent/main.go

```diff
@@ -349,3 +364,150 @@ func mainFunc() error {
+	// nb: we start looking at user id 2, because 1 should always be an unamed admin user and
+	//     we don't want to confuse that unnamed admin with an open slot we can take over.
+	exists := false
+	sideroUserID := uint8(0)
+
+	for i := uint8(2); i <= maxUsers; i++ {
+		userRes, err := ipmiClient.GetUserName(i)
+		if err != nil {
+			// nb: A failure here actually seems to mean that the user slot is unused,
+			// even though you can also have a slot with empty user as well. *scratches head*
+			// We'll take note of this spot if we haven't already found another empty one.
+			if sideroUserID == 0 {
+				sideroUserID = i
+			}
+			continue
+		}
+
+		// Found pre-existing sidero user
+		if userRes.Username == "sidero" {
+			exists = true
+			sideroUserID = i
+			log.Printf("Sidero user already present in slot %d. We'll claim it as our very own.\n", i)
+			break
+
+			// If this is the first empty user that's not the UID 1 (which we skip),
+			// we'll take this spot for sidero user
+		} else if userRes.Username == "" && sideroUserID == 0 {
+			log.Printf("Found empty user slot %d. Noting as a possible place for sidero user.\n", i)
+			sideroUserID = i
+		}
+	}
+
+	// User didn't pre-exist and there's no room
+	// Return without sidero user :(
+	if sideroUserID == 0 {
+		return errors.New("No slot available for sidero user")
+	}
+
+	// Not already present and there's an empty slot so we add sidero user
+	if !exists {
+		log.Printf("Adding sidero user to slot %d\n", sideroUserID)
+		_, err := ipmiClient.SetUserName(sideroUserID, "sidero")
+		if err != nil {
+			return err
+		}
+	}
+
+	// Reset pass for sidero user
+	// nb: we _always_ reset the user pass because we can't ever get
+	//     it back out when we find an existing sidero user.
+	pass := genPass16()
+	_, err = ipmiClient.SetUserPass(sideroUserID, pass)
+	if err != nil {
+		return err
+	}
+
+	// Make sidero an admin
+	// Options: 0xD1 == Callin false, Link false, IPMI Msg true, Channel 1
+	// Limits: 0x03 == Administrator
+	// Session: 0x00 No session limit
+	_, err = ipmiClient.SetUserAccess(0xD1, sideroUserID, 0x04, 0x00)
+	if err != nil {
+		return err
+	}
+
+	// Enable the sidero user
+	_, err = ipmiClient.EnableUser(sideroUserID)
+	if err != nil {
+		return err
+	}
+
+	// Finally fill in info for update request
+	bmcInfo.User = "sidero"
+	bmcInfo.Pass = pass
+
+	// Attempt to update server object
+	err = retry.Constant(5*time.Minute, retry.WithUnits(30*time.Second), retry.WithErrorLogging(true)).Retry(func() error {
+		ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+		defer cancel()
+
+		_, err = client.UpdateBMCInfo(
+			ctx,
+			&api.UpdateBMCInfoRequest{
+				Uuid:    uuid.String(),
+				BmcInfo: bmcInfo,
+			},
+		)
+
+		if err != nil {
+			return retry.ExpectedError(err)
+		}
+
+		return nil
+	})
+
+	return nil
+}
+
+// Returns a random pass string of len 16
+func genPass16() string {
+	rand.Seed(time.Now().UnixNano())
```

Comment:
as this function is potentially sensitive to the randomness of the data, and time might be not initialized properly (though unlikely), we might benefit from kernel level random source which should always be better, e.g. with equivalent function [crypto/rand.Int](https://pkg.go.dev/crypto/rand#Int)

Reply:
I think I've got all of your nits taken care of in amended commit. I want to look at the secret for creds first before merging though.